Unhelpful suggestions trying to use array::set

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.

If you cannot produce a minimal reproduction case (something that would work in
isolation), please provide the steps or even link to a repository that causes
the problematic output to occur.
-->

Given the following code: <!-- Please provide a link to play.rust-lang.org -->
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0eb1e294409a320c014ebbe08444c14d)
```rust
fn main() {
    let mut a = [0u8; 1];
    a.set(0, 3);
}
```

The current output is:

```
/tmp/tmp.pvMJNmiDgb/bad_error/src (git)-[main] % cargo run
   Compiling bad_error v0.1.0 (/tmp/tmp.pvMJNmiDgb/bad_error)
error[E0599]: no method named `set` found for array `[u8; 1]` in the current scope
   --> src/main.rs:3:7
    |
3   |     a.set(0, 3);
    |       ^^^
    |
   ::: /home/jonathan/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/pin.rs:659:12
    |
659 |     pub fn set(&mut self, value: P::Target)
    |            --- the method is available for `Pin<&mut [u8; 1]>` here
    |
help: consider wrapping the receiver expression with the appropriate type
    |
3   |     Pin::new(&mut a).set(0, 3);
    |     +++++++++++++  +
help: there is an associated function with a similar name
    |
3   |     a.get(0, 3);
    |       ~~~

For more information about this error, try `rustc --explain E0599`.
error: could not compile `bad_error` due to previous error
```
![image](https://user-images.githubusercontent.com/22172241/167292580-c0fa394a-a4d7-4628-9255-fc5bb82fee1a.png)


<!-- The following is not always necessary. -->

I don't think suggesting to use Pin here helps users. I tried to use set, since there's also a get so it felt natural to me that set should also exist. It may be a different issue altogether if having no set on arrays is expected, but the error message really bothered me. I tried to do a, what felt to me, trivial operation on an array, and the first suggestion is to Pin it, something which I know is useless. However, unexperienced rust users may not know this. It also doesn't solve the issue. There would be a `set` method on pin, but it wouldn't do what you expect it to do.

--- a/compiler/rustc_typeck/src/check/method/suggest.rs
+++ b/compiler/rustc_typeck/src/check/method/suggest.rs
@@ -994,6 +994,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         span,
                         rcvr_ty,
                         item_name,
+                        args.map(|args| args.len()),
                         source,
                         out_of_scope_traits,
                         &unsatisfied_predicates,
@@ -1732,6 +1733,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         span: Span,
         rcvr_ty: Ty<'tcx>,
         item_name: Ident,
+        inputs_len: Option<usize>,
         source: SelfSource<'tcx>,
         valid_out_of_scope_traits: Vec<DefId>,
         unsatisfied_predicates: &[(
@@ -1808,7 +1810,8 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         // Explicitly ignore the `Pin::as_ref()` method as `Pin` does not
                         // implement the `AsRef` trait.
                         let skip = skippable.contains(&did)
-                            || (("Pin::new" == *pre) && (sym::as_ref == item_name.name));
+                            || (("Pin::new" == *pre) && (sym::as_ref == item_name.name))
+                            || inputs_len.map_or(false, |inputs_len| pick.item.kind == ty::AssocKind::Fn && self.tcx.fn_sig(pick.item.def_id).skip_binder().inputs().len() != inputs_len);
                         // Make sure the method is defined for the *actual* receiver: we don't
                         // want to treat `Box<Self>` as a receiver if it only works because of
                         // an autoderef to `&self`

--- a/src/test/ui/suggestions/dont-suggest-pin-array-dot-set.rs
+++ b/src/test/ui/suggestions/dont-suggest-pin-array-dot-set.rs
@@ -1,0 +1,15 @@
+// https://github.com/rust-lang/rust/issues/96834
+//
+// This test case verifies that rustc does not make an unhelpful suggestion:
+//
+//     help: consider wrapping the receiver expression with the appropriate type
+//         |
+//     14  |     Pin::new(&mut a).set(0, 3);
+//         |     +++++++++++++  +
+//
+// We can tell that it isn't helpful, because `Pin::set` takes two parameters (including
+// the receiver), but the function call on line 14 supplies three.
+fn main() {
+    let mut a = [0u8; 1];
+    a.set(0, 3); //~ERROR
+}

--- a/src/test/ui/suggestions/dont-suggest-pin-array-dot-set.stderr
+++ b/src/test/ui/suggestions/dont-suggest-pin-array-dot-set.stderr
@@ -1,0 +1,9 @@
+error[E0599]: no method named `set` found for array `[u8; 1]` in the current scope
+  --> $DIR/dont-suggest-pin-array-dot-set.rs:14:7
+   |
+LL |     a.set(0, 3);
+   |       ^^^ help: there is an associated function with a similar name: `get`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0599`.

